File: aoc/2016/aoc201617.go
package main

import (
	"crypto/md5"
	"os"
)

type Path struct {
	x, y int
	path string
}

func search(passcode string) string {
	// breadth-first search
	paths := []Path{Path{0, 0, ""}}
	for {
		newpaths := []Path{}
		for _, path := range paths {
			hash := md5.Sum([]byte(passcode + path.path))
			if path.y > 0 && (hash[0]>>4) > 10 {
				newpaths = append(newpaths, Path{path.x, path.y - 1, path.path + "U"})
			}
			if path.y < 3 && (hash[0]&15) > 10 {
				if path.x == 3 && path.y == 2 {
					return path.path + "D"
				}
				newpaths = append(newpaths, Path{path.x, path.y + 1, path.path + "D"})
			}
			if path.x > 0 && (hash[1]>>4) > 10 {
				newpaths = append(newpaths, Path{path.x - 1, path.y, path.path + "L"})
			}
			if path.x < 3 && (hash[1]&15) > 10 {
				if path.x == 2 && path.y == 3 {
					return path.path + "R"
				}
				newpaths = append(newpaths, Path{path.x + 1, path.y, path.path + "R"})
			}
		}
		if len(newpaths) == 0 {
			panic("dead-ended:"+passcode)
		}
		paths = newpaths
	}
}

// depth-first search
func search2(passcode string, path Path) int {
	for {
		if path.x == 3 && path.y == 3 {
			return len(path.path)
		}
		newpaths := []Path{}
		hash := md5.Sum([]byte(passcode + path.path))
		if path.y > 0 && (hash[0]>>4) > 10 {
			newpaths = append(newpaths, Path{path.x, path.y - 1, path.path + "U"})
		}
		if path.y < 3 && (hash[0]&15) > 10 {
			newpaths = append(newpaths, Path{path.x, path.y + 1, path.path + "D"})
		}
		if path.x > 0 && (hash[1]>>4) > 10 {
			newpaths = append(newpaths, Path{path.x - 1, path.y, path.path + "L"})
		}
		if path.x < 3 && (hash[1]&15) > 10 {
			newpaths = append(newpaths, Path{path.x + 1, path.y, path.path + "R"})
		}
		if len(newpaths) == 0 {
			// dead-ended
			return 0
		}
		if len(newpaths) == 1 {
			path = newpaths[0]
			continue
		}
		longest := 0
		for _, newpath := range(newpaths) {
			l := search2(passcode, newpath)
			if l > longest {
				longest = l
			}
		}
		return longest
	}
}

func main() {
	if len(os.Args) < 2 {
		if p := search("ihgpwlah"); p != "DDRRRD" {
			panic(p)
		}
		if p := search("kglvqrro"); p != "DDUDRLRRUDRD" {
			panic(p)
		}
		if p := search("ulqzkmiv"); p != "DRURDRUDDLLDLUURRDULRLDUUDDDRR" {
			panic(p)
		}
		if p := search2("ihgpwlah", Path{0, 0, ""}); p != 370 {
			panic(p)
		}
		if p := search2("kglvqrro", Path{0, 0, ""}); p != 492 {
			panic(p)
		}
		if p := search2("ulqzkmiv", Path{0, 0, ""}); p != 830 {
			panic(p)
		}
	} else if len(os.Args) < 3 {
		println(search(os.Args[1]))
	} else {
		println(search2(os.Args[2], Path{0, 0, ""}))
	}
}

File: aoc/2024/aoc201617.hs
module AOC201617 where

import Crypto.Hash.MD5(hash)
import Data.Bits((.&.))
import Data.ByteString(ByteString,index,pack,snoc,unpack)
import qualified Data.ByteString
import Data.Char(chr,ord)
import Data.Word(Word8)

import AOC

aoc = AOC {
    day="../../2016/input/17",
    aocTests=[
        AOCTest {
            testData="ihgpwlah",
            testResult=Just $ show "DDRRRD",
            testResult2=Just "370"
            },
        AOCTest {
            testData="kglvqrro",
            testResult=Just $ show "DDUDRLRRUDRD",
            testResult2=Just "492"
            },
        AOCTest {
            testData="ulqzkmiv",
            testResult=Just $ show "DRURDRUDDLLDLUURRDULRLDUUDDDRR",
            testResult2=Just "830"
            }
        ],
    aocCode=Code {
        codeParse=parse,
        codeParse2=parse,
        codeTest=result,
        codeTest2=result2,
        codeResult=result,
        codeResult2=result2
        }
    }

parse = map (fromIntegral . ord) . filter (/= '\n')

walk :: [((Int,Int),ByteString)] -> [((Int,Int),ByteString)] -> [ByteString]
walk [] [] = []
walk nextQueue [] = walk [] nextQueue
walk nextQueue (((x,y),path):queue)
  | (x,y) == (4,4) = path : walk nextQueue queue
  | otherwise = walk enqueued queue
  where
    doors = hash path
    enqueued = enqueueU
    enqueueU
      | y > 1 && index doors 0 >= 0xb0 = ((x,y-1),snoc path 85):enqueueD
      | otherwise = enqueueD
    enqueueD
      | y < 4 && index doors 0 .&. 15 > 10 = ((x,y+1),snoc path 68):enqueueL
      | otherwise = enqueueL
    enqueueL
      | x > 1 && index doors 1 >= 0xb0 = ((x-1,y),snoc path 76):enqueueR
      | otherwise = enqueueR
    enqueueR
      | x < 4 && index doors 1 .&. 15 > 10 = ((x+1,y),snoc path 82):nextQueue
      | otherwise = nextQueue

result passcode =
    map (chr . fromIntegral) $ drop (length passcode) $ unpack $ head
                             $ walk [] [((1,1),pack passcode)]

result2 passcode =
    (+ (-length passcode)) $ Data.ByteString.length $ last
                           $ walk [] [((1,1),pack passcode)]
